### PR TITLE
feat(dynamic-forms): added move array item event

### PR DESCRIPTION
**Pull request by dereekb**
# Pull Request

## Description

Adds a move event, which can move an item's index in an Array field from one index to the next. Necessary for drag/drop interactions as the current array events can only perform one action at at time and are potentially destructive if remove is used.

## Type of Change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 Documentation update
- [ ] 🎨 Code style update (formatting, renaming)
- [ ] ♻️ Code refactoring (no functional changes)
- [ ] ⚡ Performance improvement
- [ ] ✅ Test update
- [ ] 🔧 Build/tooling update

## Changes Made

- added MoveArrayItemEvent class with fromIndex/toIndex
- added move() method to arrayEvent() builder
- moveItem() reorders resolvedItems and form value atomically
- items preserve identity (no destroy/recreate) via splice

## Testing

- [x] Unit tests added/updated
- [x] All tests pass locally
- [ ] Tested manually
- [ ] Tested edge cases

## Documentation

- [x] Documentation updated (if needed)
- [x] API documentation (TSDoc) added/updated
- [] Examples updated (if needed)

## Checklist

- [x] Code follows the [coding standards](../CODING_STANDARDS.md)
- [x] No linting errors (`pnpm lint`)
- [x] Code is formatted (`pnpm lint:fix`)
- [x] Build succeeds (`pnpm build:libs`)
- [x] Commits follow [conventional commit format](https://www.conventionalcommits.org/)
- [ ] Self-review completed
- [x] No commented-out code or console.log() statements
- [x] Type safety maintained (no `any` types)

## Summary by CodeRabbit

* **New Features**
  * Reorder/move items in form arrays atomically while preserving item identity (no recreation).

* **Documentation**
  * Added docs and examples showing array item reordering and the new move operation.

* **Tests**
  * Added extensive unit and E2E tests covering move behavior, identity preservation, edge cases, and combined move/remove/add scenarios.

### CODE REVIEW

File: packages/dynamic-forms/src/lib/fields/array/array-field.component.ts [hunk 1]
```diff
@@ -647,6 +649,54 @@ export default class ArrayFieldComponent<TModel extends Record<string, unknown>
...
 
+  /**
+   * Handles move operations — reorders an existing item without destroying it.
+   * Updates resolvedItems and form value atomically. Since the array length
+   * doesn't change, `determineDifferentialOperation` returns 'none' and no
+   * recreate is triggered. The `@for` loop tracks by `item.id`, so Angular
+   * moves the DOM node instead of destroying/recreating. The `itemPositionMap`
+   * computed auto-recomputes, propagating new indices to child linkedSignals.
+   */
+  private moveItem(fromIndex: number, toIndex: number): void {
```
Comment (antimprisacaru) [inline packages/dynamic-forms/src/lib/fields/array/array-field.component.ts:664]: One thing I noticed: items that come from the fields config aren't in `templateRegistry` — only dynamically added ones are. So if someone moves an item and later triggers a recreate (e.g., by removing a different item), the positional lookup in `performDifferentialUpdate` will pull `itemTemplates[idx]` for the new index, which is no longer the template that originally belonged to the item there.

For uniform arrays this is invisible. For the kind of heterogeneous setup the test fixture uses (different template per position), it'd resolve against the wrong one. Probably worth either stashing the template onto the item's id on first resolve, or just calling this out in the method doc so future-us doesn't get surprised.

**Review by antimprisacaru — COMMENTED** [review]
Nice find and thanks for the contribution! 

Left a small comment, otherwise just have to regen the mcp, and also since the whole motivation here is drag/drop, it'd be really nice to have an E2E scenario that actually exercises it. Beyond being a good demo, it'd give us something to lean on for the identity-preservation claims — focus staying put, dirty/touched not resetting, DOM node reuse, etc. Unit tests can't really prove those.

File: packages/dynamic-forms/src/lib/fields/array/array-field.component.ts (hunk 1 above)
Comment (dereekb) [reply packages/dynamic-forms/src/lib/fields/array/array-field.component.ts:664]: I'll work on some E2E tests and a fix for that case. I'm assuming create tests in each of the 4 UI's E2E places? I'll also make sure the heterogeneous case is mocked up properly.

Comment (antimprisacaru) [reply packages/dynamic-forms/src/lib/fields/array/array-field.component.ts:664]: I think only core tests is fine tbh. Thanks a lot :)

Comment (dereekb) [reply packages/dynamic-forms/src/lib/fields/array/array-field.component.ts:664]: I got the core E2E tests in and ran them and they pass. Working on the specific case you mentioned and will run the E2E tests locally again.

**Comment by dereekb** [comment]
Regenerated llms-full.txt.

Gave it a test run too. Works great with cdkDrag!

Screenshot 2026-04-20 at 2 51 26 PM

File: packages/dynamic-forms/src/lib/fields/array/array-field.component.ts [hunk 2]
```diff
@@ -642,6 +641,69 @@ export default class ArrayFieldComponent<TModel extends Record<string, unknown>
...
+    // the template by item ID lets the recreate path resolve the correct one.
+    const itemTemplates = this.itemTemplates();
+    const currentItems = this.resolvedItemsSignal();
+    const lo = Math.min(fromIndex, toIndex);
+    const hi = Math.max(fromIndex, toIndex);
+    for (let i = lo; i <= hi; i++) {
+      const item = currentItems[i];
+      if (item && !this.templateRegistry.has(item.id) && i < itemTemplates.length) {
+        this.templateRegistry.set(item.id, itemTemplates[i] as FieldDef<unknown>[]);
+      }
```
[the remark this answers is not in the document]
Comment (dereekb) [reply packages/dynamic-forms/src/lib/fields/array/array-field.component.ts]: Addressed in latest commit along with a new test for this case.

File: apps/e2e/primeng/src/app/testing/array-fields/array-fields.spec.ts [hunk 3]
```diff
@@ -639,4 +639,127 @@ test.describe('Array Fields E2E Tests', () => {
...
+      // Wait for values to settle
+      await expect(inputs.nth(0)).toHaveValue('Beta', { timeout: 5000 });
+      await expect(inputs.nth(1)).toHaveValue('Gamma', { timeout: 5000 });
+      await expect(inputs.nth(2)).toHaveValue('Alpha', { timeout: 5000 });
+
+      // Verify item IDs were reordered (same IDs, different positions = no destroy/recreate)
+      const movedIds = await scenario
+        .locator('.df-array-item')
+        .evaluateAll((els) => els.map((el) => el.getAttribute('data-array-item-id')));
+      expect(movedIds).toHaveLength(3);
```
[the remark this answers is not in the document]
Comment (dereekb) [reply apps/e2e/primeng/src/app/testing/array-fields/array-fields.spec.ts:679]: We'll skip this one. Claude reasoned:
                                                                                                                                                                         
  1. The existing assertions already implicitly catch null/duplicate problems. The test asserts movedIds[0] === initialIds[1], movedIds[1] === initialIds[2], movedIds[2] === initialIds[0]. If any ID were null, multiple entries would be null and these
   equality checks would incorrectly pass only in the degenerate case where ALL IDs are null — but that would also mean initialIds[1] === initialIds[0] which would make the reorder assertions trivially pass for the wrong reason. However, the test    
  also checks values (Alpha, Beta, Gamma) changed positions — so the combination of value assertions + ID assertions already provides strong coverage.                                                                                                    
  2. The data-array-item-id attribute is set by the component from a generated UUID. It would only be null if the component's rendering were fundamentally broken, which would be caught by the earlier assertions (toHaveCount(3), toHaveValue('Alpha')).
  3. Adding these assertions across all 4 adapter specs (material, bootstrap, primeng, ionic) is pure noise — it's defensive testing against a scenario that can't happen if the earlier assertions pass.                                                 
  4. No other E2E tests in the repo follow this pattern — adding it only to the move tests would be inconsistent.

**Review by antimprisacaru — APPROVED** [review]
Nicely done, thanks for the contribution!
